pkg/river/token/builder/nested_defaults_test.go

```diff
@@ -0,0 +1,229 @@
+package builder_test
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/grafana/agent/pkg/river/ast"
+	"github.com/grafana/agent/pkg/river/parser"
+	"github.com/grafana/agent/pkg/river/token/builder"
+	"github.com/grafana/agent/pkg/river/vm"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	defaultNumber      = 123
+	otherDefaultNumber = 321
+)
+
+var testCases2 = []struct {
+	name          string
+	input         interface{}
+	expectedRiver string
+}{
+	{
+		name:          "struct propagating default - input matching default",
+		input:         StructPropagatingDefault{Inner: AttrWithDefault{Number: defaultNumber}},
+		expectedRiver: "",
+	},
+	{
+		name:  "struct propagating default - input with zero-value struct",
+		input: StructPropagatingDefault{},
+		expectedRiver: `
+		inner {
+			number = 0
+		}	
+		`,
+	},
+	{
+		name:  "struct propagating default - input with non-default value",
+		input: StructPropagatingDefault{Inner: AttrWithDefault{Number: 42}},
+		expectedRiver: `
+		inner {
+			number = 42
+		}	
+		`,
+	},
+	{
+		name:          "pointer propagating default - input matching default",
+		input:         PtrPropagatingDefault{Inner: &AttrWithDefault{Number: defaultNumber}},
+		expectedRiver: "",
+	},
+	// TODO: Test fails currently: invariant violated - not clear how to explicitly set the inner block to nil in River
+	//       The builder currently encodes this into empty river block, which decodes into default value.
+	//{
+	//	name:          "pointer propagating default - nil pointer",
+	//	input:         PtrPropagatingDefault{},
+	//	expectedRiver: "",
+	//},
```

Comment:
In other words, the issue here is:
* An empty string decodes into `{Inner: &{Number: 123}}`
* But `{Inner: nil}` encodes into empty string

Not sure how we can fix it, since it's not clear how to explicitly set a block to nil in River. Maybe trying to encode a nil block should be an error?

Reply:
This is tricky because I think this is correct, but fails just because of how the tests work right now:

1. `PtrPropagatingDefault{}` _should_ encode to the empty string because the block isn't provided.
2. When decoding, defaults are applied to `PtrPropagatingDefault{}`, which sets `Inner` to `&{Number: 123}`. 
3. This means that the developer is claiming the `Inner: nil` and `Inner: &{Number: 123}` are semantically equivalent from the user's perspective; if the user provides an empty string or `inner { number = 123 }`, they decode the same way. 

I'm not sure how this should be addressed from the test scenario.